### ISSUE TITLE
How can I use aimet AutoQuant quantization encoding for QNN?

### ISSUE THREAD

**Issue by hamzeasadi** [quic/aimet#2010]
I used aimet AutoQuant for my model quantization, after that, I applied QNN. th following is the procedure that I use for quantization and QNN.

aimet quantization:

`
dev = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
paths = cfg.Paths()
MODEL_PATH = os.path.join(paths.model, 'mymodel.pt')

train_loader, train_size, ulabeld_test_loader, test_size = dst.create_loader(batch_size=10, train_percent=0.8)
one_sample_np = np.reshape(np.fromfile(open(os.path.join(paths.data, 'John_Henry_0001.raw'), 'br'), dtype='float32'), (98, 80))

one_sample_pt = torch.from_numpy(one_sample_np).unsqueeze(dim=0).unsqueeze(dim=0).float().to(dev)

def evaluate_callback(model:nn.Module,num_samples:Optional[int]=None)->float:
        auc = engine.eval_step(model=model, data=train_loader)
       return auc

def pass_calib_data(model, _):
      for img in ulabeld_test_loader:
          _ = model(img)

model = torch.load(MODEL_PATH, map_location=dev)
model.to(dev)
input_shape = (1,1,128,128)

EVAL_DATASET_SIZE = train_size
CALIBRATION_DATASET_SIZE = test_size
BATCH_SIZE = 10

auto_quant = AutoQuant(allowed_accuracy_drop=0.0,
                       unlabeled_dataset_iterable=ulabeld_test_loader,
                       default_output_bw=8, default_param_bw=8,
                       eval_callback=evaluate_callback, default_config_file=os.path.join(paths.data, 'default_config.json'))

ADAROUND_DATASET_SIZE = test_size
adaround_data_loader = ulabeld_test_loader
print(len(adaround_data_loader))
adaround_params = AdaroundParameters(adaround_data_loader, num_batches=len(adaround_data_loader), default_num_iterations=2000)
auto_quant.set_adaround_params(adaround_params)

out_dir = os.path.join(paths.output, 'all_quant')
cfg.create_dir(out_dir)
dummy_input = torch.randn(input_shape)

model, accuracy, encoding_path =\
    auto_quant.apply(model,
                     dummy_input_on_cpu=dummy_input.cpu(),
                     dummy_input_on_gpu=dummy_input.cuda(),
                     results_dir=out_dir)

encode_file = 'batchnorm_folding.encodins'
    

input_shape = (1,1,128,128)
dummy_input = torch.randn(input_shape, device=dev)
sim = QuantizationSimModel(model=model, quant_scheme=QuantScheme.post_training_tf_enhanced, dummy_input=dummy_input, default_output_bw=8, default_param_bw=8)
sim.set_and_freeze_param_encodings(encoding_path=os.path.join(paths.output,  'all_quant', '.trace', 'batchnorm_folding.encodings'))
sim.compute_encodings(forward_pass_callback=pass_calib_data, forward_pass_callback_args=None)

quat_out = sim.model(one_sample_pt)
quat_out_np = quat_out.cpu().detach().squeeze().numpy()

sim.export(path=paths.model, filename_prefix='sim_model_resnet', dummy_input=dummy_input.to('cpu'))

quat_out_np.tofile(os.path.join(paths.quantout, 'aimet_out.raw'))
`
I will get two set of quantization encodings which they are the same. one from aimet autoquant output and second one from sim.export(). 
I applied QNN as follow

`
    qnn-onnx-converter
    --arch_checker
    --input_list onnx_models/${SAMPLE_DATA}.txt
    --input_network onnx_models/${NET_NAME}.onnx
    --output_path qnn_models/${NET_NAME}_w8_a8_b8.cpp
    --batch 1
    --float_bw 32
    --weight_bw 8
    --bias_bw 8
    --act_bw 8
    --input_type "input" image 
    --input_dtype "input" float32
    --quantization_overrides onnx_models/${NET_ENCOD}.encodings
    --param_quantizer enhanced
    --act_quantizer enhanced

qnn-model-lib-generator
    -c qnn_models/${NET_NAME}_w8_a8_b8.cpp
    -b qnn_models/${NET_NAME}_w8_a8_b8.bin
    -t x86_64-linux-clang 
    -l ${NET_NAME}_w8_a8_b8.so
    -o qnn_models

qnn-context-binary-generator
    --model qnn_models/x86_64-linux-clang/lib${NET_NAME}_w8_a8_b8.so
    --backend /home/qnn_user/qnn-v2.4.1/target/x86_64-linux-clang/lib/libQnnHtp.so
    --output_dir qnn_models/context/x86_64-linux-clang
    --binary_file ${NET_NAME}_w8_a8_b8
    --log_level warm
    --config_file qnn_models/${NET_NAME}_w8_a8_b8_net.json

qnn-net-run
    --retrieve_context qnn_models/context/x86_64-linux-clang/${NET_NAME}_w8_a8_b8.bin
    --backend /home/qnn_user/qnn-v2.4.1/target/x86_64-linux-clang/lib/libQnnHtp.so
    --profiling_level basic
    --perf_profile high_performance
    --input_list onnx_models/${SAMPLE_DATA}.txt
    --input_data_type=float
    --output_data_type float_and_native
`

I get output for the QNN for the same sample that I use for aimet.
My questions:
1- Aimet's output for a single sample differs from QNN's output for the same sample. Aren't they supposed to match?

2- QNN output for the same sample but aimet AutoQuant encoding and sime.export encoding are different, although the file encodings for both are the same?

3- When we enable __input_list in qnn-onnx-converter, the output .bin file size is considerably smaller than when we disable it. Does QNN quantization apply regardless of whether --quantization_overrides is enabled or not?

**Comment by quic-akhobare** [comment]
Hi @quic-ernst. Could you please answer this question from @hamzeasadi.

**Comment by quic-ernst** [comment]
@hamzeasadi Could you please provide your question to the QNN support team? They are available via email here: [QNN Support](mailto:qnn.support@qti.qualcomm.com)

**Comment by quic-akhobare** [comment]
Thanks @quic-ernst. @hamzeasadi  - will close this issue for the present. If you have a further AIMET issue, please raise a new one.

**Comment by anuragranj** [comment]
Hi,
Has this issue been resolved? I am not able to find any tutorial/doc for going from AIMET-onnx-model + AIMET-quantization encodings to running it on an android device with QNN. The outputs simply don't match. Is there any tutorial to reproduce a standard network, say Mobilenet, with AIMET quantization + encodings on to an Android devices using QNN?

**Comment by quic-akhobare** [comment]
> Hi, Has this issue been resolved? I am not able to find any tutorial/doc for going from AIMET-onnx-model + AIMET-quantization encodings to running it on an android device with QNN. The outputs simply don't match. Is there any tutorial to reproduce a standard network, say Mobilenet, with AIMET quantization + encodings on to an Android devices using QNN?

@anuragranj  - This is a super old issue. Can you be the first person to try the "discussion" tab at the top  -https://github.com/quic/aimet/discussions and post there. We can point you to a recipe.

**Comment by anuragranj** [comment]
Thanks. Done!
